pkg/proxy/iptables/proxier.go

```diff
@@ -1066,56 +1080,73 @@ func (proxier *Proxier) syncProxyRules() {
 			activeNATChains[policyClusterChain] = true
 		}
 
-		if hasEndpoints && svcInfo.ExternallyAccessible() && svcInfo.NodeLocalExternal() {
-			if chain, ok := existingNATChains[svcXlbChain]; ok {
+		// Declare the policyLocalChain if needed.
+		if hasEndpoints && svcInfo.UsesLocalEndpoints() {
+			if chain, ok := existingNATChains[policyLocalChain]; ok {
+				proxier.natChains.WriteBytes(chain)
+			} else {
+				proxier.natChains.Write(utiliptables.MakeChainLine(policyLocalChain))
+			}
+			activeNATChains[policyLocalChain] = true
+		}
+
+		// If any "external" destinations are enabled, set up external traffic
+		// handling.  All captured traffic for all external destinations should
+		// jump to externalTrafficChain, which will handle some special-cases
+		// and then jump to externalPolicyChain.
+		if hasEndpoints && svcInfo.ExternallyAccessible() {
+			if chain, ok := existingNATChains[externalTrafficChain]; ok {
 				proxier.natChains.WriteBytes(chain)
 			} else {
-				proxier.natChains.Write(utiliptables.MakeChainLine(svcXlbChain))
+				proxier.natChains.Write(utiliptables.MakeChainLine(externalTrafficChain))
 			}
-			activeNATChains[svcXlbChain] = true
+			activeNATChains[externalTrafficChain] = true
 
-			// The XLB chain redirects all pod -> external VIP
-			// traffic to the Service's ClusterIP instead. This happens
-			// whether or not we have local endpoints; only if localDetector
-			// is implemented
-			if proxier.localDetector.IsImplemented() {
+			if !svcInfo.NodeLocalExternal() || !proxier.localDetector.IsImplemented() {
+				// If we are proxying globally (or if we can't tell local from
+				// non-local because there's no functioning local-detector), we
+				// need to masquerade, in case we cross nodes.
 				proxier.natRules.Write(
-					"-A", string(svcXlbChain),
-					"-m", "comment", "--comment",
-					`"Redirect pods trying to reach external loadbalancer VIP to clusterIP"`,
+					"-A", string(externalTrafficChain),
+					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcNameString),
+					"-j", string(KubeMarkMasqChain))
+			} else {
+				// If we are proxying only locally, we can retain the source IP
+				// in most cases.
```

Comment:
network-pod to host-network pod (in other node) requires masquerade

Reply:
I can't tell if you're saying I missed a case or just find the case weird?

Reply:
I was trying to remember, but this seems right, host-network pod in same node should not need masquerade